### PR TITLE
Generalize patching to support vendoring

### PR DESCRIPTION
**Pull request by vyasr**
This PR makes a number of significant changes to the patching infrastructure:
1. This PR reorganizes the patching logic to be based on a more principled approach. Rather than maintaining lists of patch functions that are each responsible for filtering modules to apply themselves to, patches are organized in the patches directory in a tree structure matching dask itself. Patches are found and run by importing the same relative paths within the `patches` directory corresponding to a particular dask or distributed module.
2. It adds proper support for patching submodules. Previously the loader was being disabled whenever a real dask module was being imported, but this is problematic because if some dask modules import others they will pre-populate `sys.modules` with the real modules and therefore the loader will never be used for loading a patched version of the submodule.
3. Patches are no longer just functions applied to modules, they are arbitrary functions executed when a module is imported. As a result, a wider range of modifications is possible than was previously allowed. In particular:
4. The more general functions allow the entire module import to be hijacked and redirected to other modules.
5. The new framework is used to vendor a patched version of the accessor.py module in dask, which resolves the issues observed in https://github.com/dask/dask/pull/11035.

### CODE REVIEW

File: rapids_dask_dependency/dask_loader.py [hunk 1]
```diff
@@ -3,65 +3,60 @@
...
 class DaskLoader(importlib.abc.MetaPathFinder, importlib.abc.Loader):
+    def __init__(self):
+        self._blocklist = set()
+
     def create_module(self, spec):
         if spec.name.startswith("dask") or spec.name.startswith("distributed"):
-            with self.disable(), patch_warning_stacklevel():
-                mod = importlib.import_module(spec.name)
+            with self.disable(spec.name):
+                try:
```
Comment (rjzamora) [inline rapids_dask_dependency/dask_loader.py:20]: Do you have a sense for how much overhead it is to try importing a patch for every module beginning with "dask" or "distributed"? Perhaps we should register the patches somehow so we don't even bother trying unless we expect the import to succeed? For example, maybe the `patches` directory can contain a `__init__.py` file with an `_rdd_patches = [...]` list of modules that are patched? (feel free to ignore this thought/suggestion for now - especially if the current overhead is low)

**Review by rjzamora — COMMENTED** [review]
Wow - This is impressive. I'm a bit scared of the complexity here, but I doubt there is a much simpler way to achieve what we need.

**Comment by rjzamora** [comment]
I wonder if the following sequence of events make sense?

1. Port/merge https://github.com/rapidsai/rapids-dask-dependency/pull/37 to 24.06
2. Test/review this PR a bit further (simple things are working for me locally)
3. (optional?) Remove the `__rdd_patch_accessor.py` file from this PR, since it only applies to 24.04
4. Backport all patching/testing to 24.04
5. (optional?) Add `__rdd_patch_accessor.py` file to 24.04

File: rapids_dask_dependency/patches/dask/dataframe/accessor.py [hunk 2]
```diff
@@ -0,0 +1,7 @@
+# Copyright (c) 2024, NVIDIA CORPORATION.
+
+from rapids_dask_dependency.importer import VendoredImporter
+
+# Currently vendoring this module due to https://github.com/dask/dask/pull/11035
+_importer = VendoredImporter(__name__)
+load_module = _importer.load_module
```
Comment (wence-) [inline rapids_dask_dependency/patches/dask/dataframe/accessor.py:7]: *suggestion*: Let's restrict this patching to known bad cases:

```diff
diff --git a/rapids_dask_dependency/patches/dask/dataframe/accessor.py b/rapids_dask_dependency/patches/dask/dataframe/accessor.py
index 286ccdd..b203516 100644
--- a/rapids_dask_dependency/patches/dask/dataframe/accessor.py
+++ b/rapids_dask_dependency/patches/dask/dataframe/accessor.py
@@ -1,7 +1,10 @@
 # Copyright (c) 2024, NVIDIA CORPORATION.
-
-from rapids_dask_dependency.importer import VendoredImporter
-
-# Currently vendoring this module due to https://github.com/dask/dask/pull/11035
-_importer = VendoredImporter(__name__)
-load_module = _importer.load_module
+import sys
+if sys.version_info >= (3, 11, 9):
+    from dask import __version__
+    from packaging.version import Version
+    if Version(__version__) < Version("2024.4.1"):
+        from rapids_dask_dependency.importer import VendoredImporter
+        # Currently vendoring this module due to https://github.com/dask/dask/pull/11035
+        _importer = VendoredImporter(__name__)
+        load_module = _importer.load_module
```

File: rapids_dask_dependency/dask_loader.py (hunk 1 above)
Comment (wence-) [reply rapids_dask_dependency/dask_loader.py:20]: It's lost in the noise of import.

**Review by wence- — CHANGES_REQUESTED** [review]
I like this and have only one minor suggestion to restrict the scope of patching the vendored file.

File: rapids_dask_dependency/patches/dask/dataframe/accessor.py (hunk 2 above)
Comment (rjzamora) [reply rapids_dask_dependency/patches/dask/dataframe/accessor.py:7]: Good suggestion - Thanks @wence-

File: rapids_dask_dependency/dask_loader.py (hunk 1 above)
Comment (rjzamora) [reply rapids_dask_dependency/dask_loader.py:20]: Okay, fair enough.

File: rapids_dask_dependency/patches/dask/dataframe/__rdd_patch_accessor.py
Comment (bdice) [inline rapids_dask_dependency/patches/dask/dataframe/__rdd_patch_accessor.py:1]: Does this patch need to depend on the Dask / RAPIDS version? This vendored code has changed between 2024.1.1 (the release we're pinned to for RAPIDS 24.04) and the unpinned `main` that we're using for 24.06 development. Just want to keep things straight so we know what version we're vendoring. Maybe a comment at the top of the file could indicate the git SHA or tag from which this was pulled before modification.

https://github.com/dask/dask/compare/2024.1.1..main/#diff-02020cca578450d377d1f45564123cfae3647b7f4435c55cd6b5ac41f1af2225

File: rapids_dask_dependency/patches/dask/dataframe/accessor.py (hunk 2 above)
Comment (bdice) [reply rapids_dask_dependency/patches/dask/dataframe/accessor.py:7]: Controlling when the patch is applied should also resolve my question above about differences in targeting RAPIDS 24.04 (pinned to dask 2024.1.1) and 24.06 (currently unpinned).

File: rapids_dask_dependency/dask_loader.py (hunk 1 above)
Comment (vyasr) [reply rapids_dask_dependency/dask_loader.py:20]: Yeah I did some quick measurements and it's noise. Importing dask is slow already.

File: rapids_dask_dependency/patches/dask/dataframe/__rdd_patch_accessor.py
Comment (vyasr) [reply rapids_dask_dependency/patches/dask/dataframe/__rdd_patch_accessor.py:1]: Applied Lawrence's suggestion above to resolve this.

**Review by bdice — APPROVED** [review]
The changes look good to me, now that the patch is applied only to the affected Python/Dask versions. (I have not verified that this passes locally.)

**Review by wence- — APPROVED** [review]

File: rapids_dask_dependency/patches/dask/dataframe/__rdd_patch_accessor.py [hunk 3]
```diff
@@ -0,0 +1,450 @@
...
+
+    func.__name__ = attr
+    func.__qualname__ = f"{cls.__name__}.{attr}"
+    original_prop = getattr(pd_cls, attr)
+    if isinstance(original_prop, property):
+        method = original_prop.fget
+    elif isinstance(original_prop, functools.cached_property):
+        method = original_prop.func
+    else:
+        raise TypeError("bind_property expects original class to provide a property")
```
Comment (rjzamora) [inline rapids_dask_dependency/patches/dask/dataframe/__rdd_patch_accessor.py:38]:
```suggestion
        method = original_prop
```
Minor tweak to align with the latest version of dask (just to be safe).

**Comment by raydouglass** [comment]
This PR was reverted via force push so that it can be merged via the bot instead (in #39).

**Pull request merged**